e2e/dractions/actions.go

```diff
@@ -17,26 +18,227 @@ const (
 	FiveSecondsDuration  = 5 * time.Second
 )
 
+// If AppSet/Subscription, find Placement
+// Determine DRPolicy
+// Determine preferredCluster
+// Determine PVC label selector
+// Determine KubeObjectProtection requirements if Imperative (?)
+// Create DRPC, in desired namespace
 func EnableProtection(w workloads.Workload, d deployers.Deployer) error {
 	util.Ctx.Log.Info("enter EnableProtection " + w.GetName() + "/" + d.GetName())
 
-	return nil
+	_, isSub := d.(*deployers.Subscription)
+	_, isAppSet := d.(*deployers.ApplicationSet)
+
+	if !isSub && !isAppSet {
+		return fmt.Errorf("deployer unknown")
+	}
+
+	name := GetCombinedName(d, w)
+	namespace := name
+
+	// if isAppSet {
+	// 	namespace = util.ArgocdNamespace
+	// }
+
+	drPolicyName := DefaultDRPolicyName
+	appname := w.GetAppName()
+
+	placementName := name
+	drpcName := name
+
+	placement, placementDecisionName, err := waitPlacementDecision(util.Ctx.Hub.CtrlClient, namespace, placementName)
+	if err != nil {
+		return err
+	}
+
+	util.Ctx.Log.Info("get placementdecision " + placementDecisionName)
+
+	placementDecision, err := getPlacementDecision(util.Ctx.Hub.CtrlClient, namespace, placementDecisionName)
+	if err != nil {
+		return err
+	}
+
+	clusterName := placementDecision.Status.Decisions[0].ClusterName
+	util.Ctx.Log.Info("placementdecision clusterName: " + clusterName)
+
+	// move update placement annotation after placement has been handled
+	// otherwise if we first add ocm disable annotation then it might not
+	// yet be handled by ocm and thus PlacementSatisfied=false
+	if placement.Annotations == nil {
+		placement.Annotations = make(map[string]string)
+	}
+
+	placement.Annotations[OcmSchedulingDisable] = "true"
+
+	util.Ctx.Log.Info("update placement " + placementName + " annotation")
+
+	if err = updatePlacement(util.Ctx.Hub.CtrlClient, placement); err != nil {
+		return err
+	}
+
+	util.Ctx.Log.Info("create drpc " + drpcName)
+
+	drpc := generateDRPC(name, namespace, clusterName, drPolicyName, placementName, appname)
+	if err = createDRPC(util.Ctx.Hub.CtrlClient, drpc); err != nil {
+		return err
+	}
+
+	return waitDRPCReady(util.Ctx.Hub.CtrlClient, namespace, drpcName)
 }
 
+// remove DRPC
+// update placement annotation
 func DisableProtection(w workloads.Workload, d deployers.Deployer) error {
-	util.Ctx.Log.Info("enter DisableProtection " + w.GetName() + "/" + d.GetName())
+	util.Ctx.Log.Info("enter DRActions DisableProtection")
+
+	_, isSub := d.(*deployers.Subscription)
+	_, isAppSet := d.(*deployers.ApplicationSet)
+
+	if !isSub && !isAppSet {
+		return fmt.Errorf("deployers.Deployer not known")
+	}
+
+	name := GetCombinedName(d, w)
+	namespace := name
+	placementName := name
+	drpcName := name
+	client := util.Ctx.Hub.CtrlClient
+
+	// if isAppSet {
+	// 	namespace = util.ArgocdNamespace
+	// }
 
-	return nil
+	util.Ctx.Log.Info("delete drpc " + drpcName)
+
+	if err := deleteDRPC(client, namespace, drpcName); err != nil {
```

Comment:
We would need to wait for DRPC to be deleted before taking next step here, as DRPC deletion can take time to cleanup the required VRG and such from the managed clusters.

I would suggest addition of a generic waitForDeletion here and then proceed.

Reply:
DRPC deletion action will not be completed util all the VRG and others resources are processed by the finalizer. So this already ensures what you want. Agree?

Reply:
If I understand correctly you are saying that `ctrlClient.Delete` in `deleteDRPC` will not return before the resource is actually deleted from the API server? What I understood was that it would mark the resource as deleted (IOW add the deletion TS etc.) but return once that is done without waiting for the DRPC to be actually deleted.

What am I missing?

Reply:
Shyam is correct, the controller runtime delete func will return immediately and not be blocked. The resource enters a "terminating" state. The deletionTimestamp is set, and the resource is marked for deletion but is not completely removed until all finalizers are handled.